lib/job_manager/job.go

```diff
@@ -0,0 +1,213 @@
+package jobmanager
+
+import (
+	"context"
+	"errors"
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"syscall"
+)
+
+type Job struct {
+	ID          string
+	LogBuffer   [][]byte
+	LogChannels []chan []byte
+	DoneChannel chan struct{}
+
+	status JobStatus
+	mu     sync.Mutex
+	cond   *sync.Cond
+	wg     sync.WaitGroup
+
+	cmd *exec.Cmd
+}
+
+func NewJob(id string) *Job {
+	job := &Job{
+		ID:          id,
+		LogBuffer:   make([][]byte, 0),
+		LogChannels: make([]chan []byte, 0),
+		DoneChannel: make(chan struct{}),
+		status:      JobStatusInitializing,
+	}
+	job.cond = sync.NewCond(&job.mu)
+	return job
+}
+
+func (j *Job) Lock() {
+	j.mu.Lock()
+}
+
+func (j *Job) Unlock() {
+	j.mu.Unlock()
+	j.cond.Broadcast()
+}
+
+func (j *Job) SetStatus(status JobStatus) {
+	j.Lock()
+	defer j.Unlock()
+	
+	j.status = status
+}
+
+func (j *Job) GetStatus() JobStatus {
+	j.Lock()
+	defer j.Unlock()
+
+	return j.status
+}
+
+func (j *Job) Start(ctx context.Context, command string, args ...string) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				j.SetDone(JobStatusError)
+			}
+		}()
+		j.cmd = exec.CommandContext(ctx, command, args...)
+		stdout, err := j.cmd.StdoutPipe()
+		if err != nil {
+			log.Fatalf("Failed to get stdout pipe: %v", err)
+		}
+
+		if err := j.cmd.Start(); err != nil {
+			log.Fatalf("Failed to start command: %v", err)
+		}
+
+		// Update job status
+		log.Print("Setting job status to running")
+		j.SetStatus(JobStatusRunning)
+
+		// Log command output
+		log.Print("Logging command output")
+		j.wg.Add(1)
+		go j.logOutput(stdout)
+
+		// Wait for the command to finish
+		if err := j.cmd.Wait(); err != nil {
+			// Update Job Status
+			log.Print("Setting job status to error")
+			j.SetStatus(JobStatusError)
+
+			log.Printf("Command failed: %v", err)
+		}
+
+		j.Lock()
+		defer j.Unlock()
+
+		log.Print("Closing job")
+		j.SetDone(JobStatusDone)
+	}()
+}
+
+func (j *Job) SetDone(status JobStatus) {
```

Comment:
being this publicly exposed, a user can call it multiple times which will cause panics because multiple channel close operations